src/dune_rules/melange_rules.ml

```diff
@@ -27,23 +49,43 @@ let js_includes ~loc ~sctx ~target_dir ~requires_link ~scope ~js_ext =
         let project = Scope.project scope in
         let deps_of_lib (lib : Lib.t) =
           let lib_name = Lib.name lib in
-          let lib = local_of_lib ~loc lib in
-          let info = Lib.Local.info lib in
-          let lib_dir = Lib_info.src_dir info in
-          let dst_dir = lib_output_dir ~target_dir ~lib_dir in
-          let open Memo.O in
-          let modules_group =
-            Dir_contents.get sctx ~dir:lib_dir
-            >>= Dir_contents.ocaml
-            >>| Ml_sources.modules ~for_:(Library lib_name)
-          in
-          let* source_modules = modules_group >>| Modules.impl_only in
-          let of_module m =
-            let output = make_js_name ~js_ext ~dst_dir m in
-            Dep.file (Path.build output)
+          let info =
+            let lib = local_of_lib ~loc lib in
+            Lib.Local.info lib
           in
-          Resolve.Memo.return
-            (List.map source_modules ~f:of_module |> Dep.Set.of_list)
+          match Lib_info.virtual_ info with
+          | Some _ ->
+            (* Js rules for virtual libs are added by their implementations *)
+            Resolve.Memo.return ([] |> Dep.Set.of_list)
+          | None ->
+            let lib_dir = Lib_info.src_dir info in
+            let dst_dir = lib_output_dir ~target_dir ~lib_dir in
+            let open Memo.O in
+            let modules_group =
+              Dir_contents.get sctx ~dir:lib_dir
+              >>= Dir_contents.ocaml
+              >>| Ml_sources.modules ~for_:(Library lib_name)
+            in
+            let* source_modules = modules_group >>| Modules.impl_only in
```

Comment:
This isn't quite right I believe. `impl_only` returns all the modules with implementations which will include modules from  the vlib and impl. This means you'll set up the same rules twice. This is how it should go:

1. Get the modules for a library using `Dir_contents.modules_of_lib`.
2. Use `Modules.split_by_lib` to get the modules for the virtual library and implementation separately
3. Setup rules for every individual module if `Module.has ~ml_kind:Impl` returns true.